### ISSUE TITLE
ComputeV2 fails when receiving multiple endpoints

### ISSUE THREAD

**Issue by ismacasti** [gophercloud/gophercloud#1764]
When more than one endpoint is configured for Nova, `NewComputeV2` fails with _Discovered 2 matching endpoints_.

The following comment can be found in openstack/endpoint_location.go
> It's an error both when multiple endpoints match the provided
criteria and when none do. 

I can't find anywhere in OpenStack API documentation that having more than one endpoint is not correct. Python OS client handles it without dying.

Example code, logs and endpoint list here in this gist:
https://gist.github.com/ismacasti/9086211b83596e6a50268543ece54d05

Thanks!

**Comment by jtopjian** [comment]
Hi, @ismacasti 

From looking at the info you gave, it appears your cloud has multiple entries to support an IPv4 and IPv6 endpoint. I can see the reasoning for doing this, and while the docs don't explicitly say that having duplicate entries is wrong, I also cannot find any evidence that configuring keystone to have an IPv4 and IPv6 endpoint is a valid configuration.

I've always assumed that if the cloud wanted to support both IPv4 and IPv6 endpoints, the URLs should be modified to use host/domain names instead of IP addresses. This way DNS can naturally handle IPv4 or IPv6 resolution.

I spent some time digging through the Python code in order to understand why this error doesn't occur. The behavior you see in Python is due to this:

https://github.com/openstack/keystoneauth/blob/master/keystoneauth1/access/service_catalog.py#L435-L436

If more than one endpoint is found, only the first will be returned.

I validated this by creating multiple endpoints in a catalog and found that the returned endpoint is the one which is the first result when sorted by endpoint ID. With this in mind, is it possible that the second endpoint in your cloud isn't being used at all?

Back to Gophercloud: you should be able to work around this problem in Gophercloud by manually creating a service client like so:

```go
serviceClient := new(gophercloud.ServiceClient)
serviceClient.ProviderClient = providerClient
serviceClient.Endpoint = "http://1.2.3.4:8774/v2.1/"
serviceClient.Type = "compute"
```

This means that your application using Gophercloud will need to manually create service clients rather than using the other convenience functions.

I'll think about whether this should be "fixed" within Gophercloud. I'm not a fan of altering the code to account for non-standard OpenStack environments (this is the first reported case of this issue in the 3+ years I've worked with Gophercloud), but on the other hand, there are benefits to mirroring the behavior of the Python libraries. 

Please let me know if you have any questions.

**Comment by jtopjian** [comment]
> Mostly it's random, as the order of endpoints is not stable.

What version of Keystone are you running in production? I can confirm that as of Stein, the result is a consistently returned endpoint with the other endpoint never being used. Just something to keep in mind.

> This behaviour would be my preferred option, it follows the Python SDK style and it allows this specific case to at least work.

Well, yes, I'd assume the preferred option is the one which doesn't require modifying the OpenStack environment 😉 

I'll try to find some time soon to investigate this change.

**Comment by ismacasti** [comment]
Hi @jtopjian,

I'm not 100% sure, but I think it's Pike. At least for us the order doesn't seem static.

**Comment by jtopjian** [comment]
Implemented in #1766
